### PR TITLE
Calculation of SCF expansion based on an N-body representation of the density

### PR DESCRIPTION
**Pull request by morganb-phys**
Introduce SCF expansion for n-body simulations. Added two new functions in SCFPotential.py for calculating SCF coefficients from an n-body simulation:
-scf_compute_coeffs_spherical_nbody
-scf_compute_coeffs_nbody

Also include a test for the spherical case. Non-spherical test to be added soon.

### CODE REVIEW

**Comment by jobovy** [comment]
Can you put the tests just in `test_scf`? I think it fails because `HernquistPotential` isn't imported.

**Comment by morganb-phys** [comment]
I've deleted test_SCFPotential.py and added the test to test_scf.py

File: galpy/potential/SCFPotential.py [hunk 1]
```diff
@@ -685,6 +714,108 @@ def integrand(xi, costheta):
...
+
+        """ 
+        
+        n = numpy.arange(0,N)
+        l = numpy.arange(0,L)
+        m = numpy.arange(0,L)
+
+        r = numpy.sqrt(pos[0]**2+pos[1]**2+pos[2]**2)
+        phi = numpy.arctan2(pos[1],pos[0])
+        costheta = pos[2]/r
```
Comment (nstarman) [inline galpy/potential/SCFPotential.py:763]: Astropy does not (yet) support the ufunc ``lpmv``, so dimensionless quantities need to be recast.
The following works on both normal and astropy Quantity arrays.

``costheta = u.Quantity(pos[2]/r, copy=False).to_value(u.one)``

File: galpy/potential/SCFPotential.py [hunk 2]
```diff
@@ -685,6 +714,108 @@ def integrand(xi, costheta):
...
+        sinmphi= numpy.sin(phi[:,None]*m[None,:])[:,None,None,:]
+        
+        Ylm= (numpy.sqrt((2.*l[:,None]+1)*gamma(l[:,None]-m[None,:]+1)/gamma(l[:,None]+m[None,:]+1))[None,None,:,:]*Plm)[None,:,:,:,:]*numpy.array([cosmphi,sinmphi])
+        Ylm= numpy.nan_to_num(Ylm)
+
+        C= [[gegenbauer(nn,2.*ll+1.5) for ll in l] for nn in n] 
+        Cn= numpy.zeros((1,len(r),N,L,1))
+        for i in range(N):
+            for j in range(L):
+                Cn[0,:,i,j,0]= C[i][j]((r/a-1)/(r/a+1))
```
Comment (nstarman) [inline galpy/potential/SCFPotential.py:780]: Likewise, ``eval_ gegenbauer`` is not an astropy-supported ufunc, so the following is necessary.

``Cn[0,:,i,j,0]= C[i][j]( u.Quantity((r/a-1)/(r/a+1), copy=False).to_value(u.one) )``

File: galpy/potential/SCFPotential.py [hunk 3]
```diff
@@ -685,6 +714,108 @@ def integrand(xi, costheta):
...
+                    Plm= lpmv(mm,ll,costheta)
+
+                    cosmphi= numpy.cos(phi*mm)
+                    sinmphi= numpy.sin(phi*mm)
+
+                    Ylm= (numpy.sqrt((2.*ll+1)*gamma(ll-mm+1)/gamma(ll+mm+1))*Plm)[None,:]*numpy.array([cosmphi,sinmphi])
+                    Ylm= numpy.nan_to_num(Ylm)
+
+                    C= gegenbauer(nn,2.*ll+1.5)
+                    Cn= C((r/a-1)/(r/a+1))
```
Comment (nstarman) [inline galpy/potential/SCFPotential.py:807]: Likewise, ``eval_ gegenbauer`` is not an astropy-supported ufunc, so the following is necessary.

``Cn = C( u.Quantity((r/a-1)/(r/a+1), copy=False).to_value(u.one) )``

But there's a computational problem if *r* is close to 0. The argument simplifies to

``Cn= C( u.Quantity((a+1)/(a-1), copy=False).to_value(u.one) )``

File: galpy/potential/SCFPotential.py [hunk 4]
```diff
@@ -685,6 +714,108 @@ def integrand(xi, costheta):
...
+
+        Inl= (-Knl*4.*numpy.pi/2.**(8.*l[None,:]+6.)*gamma(n[:,None]+4.*l[None,:]+3.)/gamma(n[:,None]+1)/(n[:,None]+2.*l[None,:]+1.5)/gamma(2.*l[None,:]+1.5)**2)[None,:,:,None]
+
+        Anlm= Inl**(-1)*Sum'''
+        Anlm= numpy.zeros([2,L,L,L])
+        for i,nn in enumerate(n):
+            for j,ll in enumerate(l):
+                for k,mm in enumerate(m[:j+1]):
+
+                    Plm= lpmv(mm,ll,costheta)
```
Comment (nstarman) [inline galpy/potential/SCFPotential.py:798]: Astropy does not (yet) support the ufunc ``lpmv``, so dimensionless quantities need to be recast.
The following works on both normal and astropy Quantity arrays.

``costheta = u.Quantity(pos[2]/r, copy=False).to_value(u.one)``

File: galpy/potential/SCFPotential.py (hunk 1 above)
Comment (jobovy) [reply galpy/potential/SCFPotential.py:763]: To deal with positions given with units, we'll just have to apply the relevant decorator. I think probably the one used for the 2D potential functions should work?

**Pull request merged**